relaynet-cogrpc-jvm/src/main/java/tech/relaycorp/relaynet/cogrpc/client/CogRPCClient.kt

```diff
@@ -1,19 +1,132 @@
 package tech.relaycorp.relaynet.cogrpc.client
 
+import io.grpc.netty.NettyChannelBuilder
+import io.grpc.stub.StreamObserver
+import kotlinx.coroutines.channels.BroadcastChannel
+import kotlinx.coroutines.channels.sendBlocking
 import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.asFlow
+import kotlinx.coroutines.flow.channelFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.emptyFlow
+import kotlinx.coroutines.flow.onCompletion
+import kotlinx.coroutines.flow.onEach
+import tech.relaycorp.relaynet.cogrpc.Authorization
+import tech.relaycorp.relaynet.cogrpc.CargoDelivery
+import tech.relaycorp.relaynet.cogrpc.CargoDeliveryAck
+import tech.relaycorp.relaynet.cogrpc.CargoRelayGrpc
 import tech.relaycorp.relaynet.cogrpc.CogRPC
+import java.net.InetSocketAddress
+import java.net.URL
+import java.util.logging.Level
+import java.util.logging.Logger
 
-interface CogRPCClient {
+class CogRPCClient(
+    serverAddress: String,
+    useTls: Boolean = true
+) {
+
+    private val channel by lazy {
+        val url = URL(serverAddress)
+        val address = InetSocketAddress(
+            url.host,
+            url.port.let { if (it != -1) it else DEFAULT_PORT }
+        )
+        NettyChannelBuilder
+            .forAddress(address)
+            .useTransportSecurity()
+            .run { if (useTls) useTransportSecurity() else usePlaintext() }
+            .build()
+    }
 
     @Throws(Exception::class)
-    fun deliverCargo(cargoes: Iterable<CogRPC.MessageDelivery>): Flow<CogRPC.MessageDeliveryAck>
+    fun deliverCargo(cargoes: Iterable<CogRPC.MessageDelivery>): Flow<CogRPC.MessageDeliveryAck> {
```

Comment:
It seems `none()` does not load everything, just checks `isEmpty()` or `hasNext()`. `map` does cause all objects to be loaded but:
- Although it's an `Iterable`, we already had them in memory before, so there's no extra harm here
- More importantly, we're passing around InputStream, and not the actual bytes. Since they are FileInputStreams underneath, the files are only actually read on the `forEach` below, when sending them to the server.

Reply:
Thanks for clarifying that. Especially the bit about `InputStream`, I forgot about that. Let's keep this as it is in that case.